### DISCUSSION TITLE
Pin number confusion and SPI fail

### DISCUSSION THREAD

**Discussion by strud** [jgromes/RadioLib#446] [Q&A] [answered]
Iam implementing a system using an ESP32 WROOM module + a NiceRF LoRa126XF30 Module. This LoRa module is the 433MHz version and is using an SX1268.
The std SPI interface is provided, along with DIO1, DIO3, NRESET and BUSY pins.

My current issue lies with the SPI interface not working really at all and I think it is due to some level of confusion I have with what is shown in the examples for the module configuration arguments.

From the example:

```c++
// SX12628 has different connections:
// NSS pin:   8
// DIO1 pin:  4
// NRST pin:  5
// BUSY pin:  6
SX1268 radio2 = new Module(8, 4, 5, 6);
```

Q: What are these pin numbers on? Is this for the default hardware?

For my ESP32 I am using the default SPI interface (VSPI) hence my initialisation looks like this:

```c++
SX1268 radioModem = new Module(VCS,DIO1, NRESET,BUSY, SPI, RADIOLIB_DEFAULT_SPI_SETTINGS);
```

where VCS = 5, DIO1 = 35, NRESET=14, BUSY=27.....
My SPI pins are per default VSPI ie CS=5, MISO = 19, MOSI = 23, CLK = 18

When the following is called, very odd SPI data is captured on the logic analyser that doesn't look like the clock is running, that MOSI is a flatline and that MISO is responding as a mirror to CS.

At the moment, the initialisation attempt returns a failure code of 2

```
ie:  int radioState = radioModem.begin(430.0,500.0,10,5,0x12,22,20,1.6,false);
```

radiostate returned: -2

explanation:
 \brief Radio chip was not found during initialization. This can be caused by specifying wrong chip type in the constructor
  (i.e. calling SX1272 constructor for SX1278 chip) or by a fault in your wiring (incorrect slave select pin).
*/
#define RADIOLIB_ERR_CHIP_NOT_FOUND                            (-2)

Any suggestions most appreciated.

Craig

------------------------------------------------------------------------------------------------------------------------------------------------------
Ok since the original post, I have tried starting the SPI in my code prior to calling the begin function and this has made a big difference.

New line:

SPI.begin();

So the difference is the SPI bus is active (wahoo!) and the module seems to be responding (MISO responds) but ALL of the responses are 0xA2.

![image](https://user-images.githubusercontent.com/66097047/150629992-da7695ce-a594-47f3-ada9-db42e0dc3e1e.png)

So in the end, I still have an issue, that being the begin function fails with error code -707

/*!
  \brief SX126x failed to execute SPI command.
*/
#define RADIOLIB_ERR_SPI_CMD_FAILED                            (-707)

Thoughts?

**Comment by jgromes** [comment, marked as answer]
Since you provided an SPI instance in the Module constructor, RadioLib assumed that you don't want to automatically initialize the interface (it's documented in the API docs). If you don't provide the SPI and SPI settings, RadioLib will use the defaults and autoamtically initialize the SPI for you.

Next, I would suggest running the code in debug verbose mode, to verify that what you see on the MISO line is what the ESP32 is reading. 

You should also check whther the module you're using has a crystal, or TCXO. Since it has DIO3 accessible, I would guess it has just plain crystal. However, you're controlling it as if it had a TCXO by providing the reference voltage of 1.6 V in the begin method. Set it to 0 to disable it.

**Reply by strud** [reply]
Hi Jan

Many thanks for your considered and useful answer. Also for such an awesome library! both in capability and the quality of the code AND the documentation.

Your suggestion about he crystal was spot on. After setting the reference voltate to 0, the device lives and now responds to commands. 
I will bear this in mind when using their 915MHz model, as this uses a TXCO instead...

The debug verbose mode is very useful indeed.

Thanks again!

**Reply by jgromes** [reply]
I'm glad it's working for you now! I've actually seen this sort of problem before, I think I'll add a note the error code. When using an XTAL module as if it had TCXO, it always returns SPI command fail. If using TCXO as if it had an XTAL, it ends up in infinite loop during calibration.

**Comment by almarion** [comment]
Hi,
I have a couple of 2W LoRa modules ( Lora1262F30-TXCO-433) from NiceRF spi interfaced to Arduino Nano 33 BLE boards. They seem to work fine and will communicate with other of my LoRa boards. I am using
https://github.com/StuartsProjects/SX12XX-LoRa/tree/master/examples/SX126x_examples/Basics for the code.

I have used RadioLIb successfully in the past, but only on other projects so far.
Anyway my main problem now is how to actually configure the 2 W modules for the complete range of Tx power up to 2 W (32 dBm).
Can you provide some guidance in this?
Thanks,
Alan

PowerlevelofDEMO board 0 1 2 3 4 5 6 7 8 9
Registervalue -5-2 1 4 7 10 13 16 19 22
Power(dBm) 13.6 16.1 18.9 21.6 24.1 25.3 27.8 30.3 31.4 32

**Reply by strud** [reply]
Hi Alan,

Here is the code I used during the initialisation and it seems to work well for me.

Note a power setting of 22 is full power ie 33dBm, but this is only available at 6V supply from memory.

`//--------------------------------------------------------------
// Initialise Radio

// initialize the second LoRa instance with
  // non-default settings
  // this LoRa link will have high data rate,
  // but lower range
  // carrier frequency:           430.0 MHz
  // bandwidth:                   500.0 kHz
  // spreading factor:            10
  // coding rate:                 5
  // sync word:                   0x34 (public network/LoRaWAN)
  // output power:                22 -> 33dBm
  // preamble length:             20 symbols
  
  
  Serial.print(F("[SX1268] Initializing"));
  digitalWrite(ARM_STATUS, 1);

  //radioState = radioModem.begin(430.0F,250.0, 10, 5, 0x34, 22,20,);
  radioState = radioModem.begin(430.0F,250.0, 10, 5, 0x34, 22,20,0,false);
  digitalWrite(ARM_STATUS, 0);
  if (radioState == RADIOLIB_ERR_NONE) {
      Serial.println(F("success!"));
      radioConfigure(&localSystem.radioModemSettings, &radioModem);
    } else {
      Serial.print(F("failed, code "));
      Serial.println(radioState);
      //while (true);
   }`

**Comment by almarion** [comment]
Hi Strud,
Thank you for giving me some hope in enabling 2W.  Are you sure that you are actually getting 30+ dBm?  So far I have only been able to get the max 22 dBm.  I know my modules are capable of producing the full 2W because I have demo kits that make this easy to see and in my case monitor with SDR.  It seems to me that there must be some SX1262 register that is being used to enable a PA or some such and that this not addressed in all the Arduino apps I have tried.  See this from RadioLib:
https://github.com/jgromes/RadioLib/wiki/High-power-Radio-Modules-Guide
G-Nice RF 1268F30
SX1268 LoRa module
unknown PA
Really appreciate any insight you may have,
Alan VA3IHS /VA7
